### PR TITLE
fix(storage-proofs): spec sync 

### PR DESCRIPTION
**Pull request by dignifiedquire**
The first round of updating naming and adjusting some smaller things better matching the spec.

Important to note
- fixed logic in tapering, such that it starts at the last layer note the first
- split labeling proofs and encoding proofs
- add missing round of labeling proofs for the last layer

### CODE REVIEW

File: storage-proofs/src/circuit/stacked/labeling_proof.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        )?;
+
+        // enforce equality
+        constraint::equal(&mut cs, || "equality_key", &exp_encoded_node, &key);
+
+        Ok(())
+    }
+}
+
+impl<H: Hasher> From<VanillaLabelingProof<H>> for LabelingProof {
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/stacked/labeling_proof.rs]: why not use vanillaproof in the spec instead?

File: storage-proofs/src/stacked/challenges.rs [hunk 2]
```diff
@@ -30,7 +30,7 @@ impl LayerChallenges {
         assert!(layer <= self.layers, "Layer too large");
 
         // TODO: proper tapering
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/challenges.rs]: ok

File: storage-proofs/src/stacked/labeling_proof.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        hasher.update(&(self.node as u64).to_le_bytes());
+
+        for parent in &self.parents {
+            hasher.update(AsRef::<[u8]>::as_ref(parent));
+        }
+
+        bytes_into_fr_repr_safe(hasher.finalize().as_ref()).into()
+    }
+
+    pub fn verify(&self, replica_id: &H::Domain, exp_labeled_node: &H::Domain) -> bool {
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/labeling_proof.rs]: expected

File: storage-proofs/src/stacked/params.rs [hunk 4]
```diff
@@ -201,46 +207,43 @@ impl<H: Hasher, G: Hasher> Proof<H, G> {
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: You are right, there were some checks missing.

File: storage-proofs/src/circuit/stacked/params.rs [hunk 5]
```diff
@@ -78,36 +85,30 @@ impl<H: Hasher, G: Hasher> Proof<H, G> {
...
-                proof.synthesize_key(
-                    cs.namespace(|| format!("encoding_proof_{}", layer)),
-                    params,
-                    replica_id,
-                    &encoded_node,
-                )?;
-            }
+        for (layer, proof) in labeling_proofs.into_iter() {
+            let raw = replica_column_proof.c_x.get_node_at_layer(layer);
+            let encoded_node =
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/stacked/params.rs]: fixed

File: storage-proofs/src/circuit/stacked/labeling_proof.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        } else {
+            params.graph.degree()
+        };
+        LabelingProof {
+            node: None,
+            parents: vec![None; degree],
+        }
+    }
+
+    fn create_key<CS: ConstraintSystem<Bls12>>(
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/stacked/labeling_proof.rs]: fixed

File: storage-proofs/src/circuit/stacked/labeling_proof.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                while bits.len() % 8 > 0 {
+                    bits.push(Boolean::Constant(false));
+                }
+                Ok(bits)
+            })
+            .collect::<Result<Vec<_>, SynthesisError>>()?;
+
+        let node_num = uint64::UInt64::alloc(cs.namespace(|| "node"), node)?;
+
+        kdf(
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/stacked/labeling_proof.rs]: fixed

File: storage-proofs/src/stacked/params.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: added them

File: storage-proofs/src/stacked/params.rs [hunk 8]
```diff
@@ -327,49 +330,49 @@ pub struct PersistentAux<D> {
 #[derive(Debug)]
 pub struct TemporaryAux<H: Hasher, G: Hasher> {
     /// The encoded nodes for 1..layers.
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: I believe this is being checked now

File: storage-proofs/src/stacked/params.rs [hunk 9]
```diff
@@ -201,46 +207,43 @@ impl<H: Hasher, G: Hasher> Proof<H, G> {
...
-                    self.comm_r_last_proof.leaf(),
-                    Some(self.comm_d_proofs.leaf()),
-                )
-            } else {
-                (
-                    self.replica_column_proofs.c_x.get_node_at_layer(layer),
-                    None,
-                )
-            };
+            let labeled_node = self.replica_column_proofs.c_x.get_node_at_layer(layer);
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: fixed

File: storage-proofs/src/stacked/params.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: fixed the `comm_r` part

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: >  we also need to check inclusion-proof path length 

why?

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: to be more specific, wouldn't finding a different path with a different length, that hashes to the same root be equivalent to a hash collision?

File: storage-proofs/src/circuit/stacked/labeling_proof.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/circuit/stacked/labeling_proof.rs]: It's kind of an opaque legacy name in the implementation, and doesn't really give any insight into the structure of the proofs. I was hoping to improve on that.

File: storage-proofs/src/stacked/params.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/stacked/params.rs]: No, there is an edge case. If the challenge being checked represents the path bits implicitly as an integer, then leading zeros are not considered. The clearest example is a challenge to the first node. A fake proof with only one hash (the right hash of the merkle root) could be provided and would verify as valid. In fact, the root itself could masquerade as zero-length inclusion proof of the zeroth node of a zero-height tree built on data of size one node.

In practice, such fake proofs might not pose much problem, but they are an example of why if you don't somehow check height/path-length, then fake proofs are possible.

File: storage-proofs/src/stacked/params.rs [hunk 10]
```diff
@@ -181,20 +176,11 @@ impl<H: Hasher, G: Hasher> Proof<H, G> {
...
-        // just grabbing the first one
-        let actual_comm_r = self.comm_r();
-        let expected_comm_r = if let Some(ref tau) = pub_inputs.tau {
-            &tau.comm_r
-        } else {
-            return false;
-        };
-
-        check_eq!(expected_comm_r, &actual_comm_r);
-
```
Comment (porcuquine) [inline storage-proofs/src/stacked/params.rs]: Good that you no longer recalculate `comm_r`, but I think you still need to verify that that any `comm_c` or `comm_r_last` used in the individual proofs are actually the same as the ones used to verify `comm_r` once at the beginning.

**Pull request merged**